File: app/src/main/java/com/karol/sezonnazdrowie/view/fragments/SettingsFragment.java
package com.karol.sezonnazdrowie.view.fragments;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.karol.sezonnazdrowie.R;
import com.karol.sezonnazdrowie.model.MainViewModel;
import com.karol.sezonnazdrowie.model.SnzAlarmManager;
import com.karol.sezonnazdrowie.view.MainActivity;
import com.karol.sezonnazdrowie.view.controls.TimePreference;
import com.karol.sezonnazdrowie.view.controls.TimePreferenceDialogFragmentCompat;

import java.util.Set;

import androidx.fragment.app.DialogFragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.navigation.Navigation;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

public class SettingsFragment extends PreferenceFragmentCompat
        implements SharedPreferences.OnSharedPreferenceChangeListener, Preference.OnPreferenceClickListener {

    private static final String TAG = "SETTINGSFRAGMENT";
    private MainViewModel mMainViewModel;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mMainViewModel = ViewModelProviders.of(getActivity()).get(MainViewModel.class);
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.prefs, rootKey);

        setPreferenceSummary(getPreferenceManager().getSharedPreferences(), "pref_season_start");
        setPreferenceSummary(getPreferenceManager().getSharedPreferences(), "pref_season_end");
        setTimePreferenceSummary(getPreferenceManager().getSharedPreferences(), "pref_notification_hour");

        findPreference("pref_notification_fruit").setOnPreferenceClickListener(this);
        findPreference("pref_notification_vegetable").setOnPreferenceClickListener(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        mMainViewModel.setActionBarTitle(getString(R.string.settings));
        return super.onCreateView(inflater, container, savedInstanceState);
    }

    @Override
    public void onResume() {
        super.onResume();
        getPreferenceManager().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
        if (mMainViewModel.isSettingsItemChanged()) {
            SnzAlarmManager.startSetAlarmsTask(getActivity(), mMainViewModel.getDatabase());
        }
        mMainViewModel.setSettingsItemChanged(false);
    }

    @Override
    public void onPause() {
        getPreferenceManager().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
        super.onPause();
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        Log.d(TAG, "onSharedPreferenceChanged: key = " + key);
        switch (key) {
            case "pref_season_start":
                setPreferenceSummary(sharedPreferences, key);
                break;
            case "pref_season_end":
                setPreferenceSummary(sharedPreferences, key);
                break;
            case "pref_notification_hour":
                setTimePreferenceSummary(sharedPreferences, key);
                break;
        }

        if (!key.equals("maxReqCode")) {
            SnzAlarmManager.startSetAlarmsTask(getActivity(), mMainViewModel.getDatabase());
        }
    }

    private void setPreferenceSummary(SharedPreferences sharedPreferences, String key) {
        Set<String> stringSet = sharedPreferences.getStringSet(key, null);
        if (stringSet == null || stringSet.size() == 0) {
            findPreference(key).setSummary(R.string.no_notification);
        } else {
            StringBuilder builder = new StringBuilder();
            for (String str : stringSet) {
                if (builder.length() != 0)
                    builder.append(", ");
                builder.append(str);
            }
            findPreference(key).setSummary(builder.toString());
        }
    }

    private void setTimePreferenceSummary(SharedPreferences sharedPreferences, String key) {
        String hourStr = sharedPreferences.getString(key, null);
        if (hourStr == null) {
            findPreference(key).setSummary("20:00");
        } else {
            findPreference(key).setSummary(hourStr);
        }
    }

    @Override
    public boolean onPreferenceClick(Preference preference) {
        Bundle bundle = new Bundle();
        if (preference.getKey().equals("pref_notification_fruit")) {
            bundle.putString(MainActivity.INTENT_WHAT, MainActivity.INTENT_WHAT_FRUITS);
        } else if (preference.getKey().equals("pref_notification_vegetable")) {
            bundle.putString(MainActivity.INTENT_WHAT, MainActivity.INTENT_WHAT_VEGETABLES);
        }

        Navigation
                .findNavController(getActivity(), R.id.nav_host_fragment)
                .navigate(R.id.settingsItemsFragment2, bundle);

        return true;
    }

    @Override
    public void onDisplayPreferenceDialog(Preference preference) {
        DialogFragment dialogFragment = null;
        if (preference instanceof TimePreference) {
            dialogFragment = new TimePreferenceDialogFragmentCompat();
            Bundle bundle = new Bundle(1);
            bundle.putString("key", preference.getKey());
            dialogFragment.setArguments(bundle);
        }

        if (dialogFragment != null) {
            dialogFragment.setTargetFragment(this, 0);
            dialogFragment.show(this.getFragmentManager(), "DIALOG");
        } else {
            super.onDisplayPreferenceDialog(preference);
        }
    }
}

File: app/src/main/java/com/karol/sezonnazdrowie/view/fragments/SettingsFragment.kt
package com.karol.sezonnazdrowie.view.fragments

import android.content.SharedPreferences
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.fragment.app.viewModels
import androidx.navigation.Navigation
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.karol.sezonnazdrowie.R
import com.karol.sezonnazdrowie.model.MainViewModel
import com.karol.sezonnazdrowie.model.PREF_NOTIFICATION_FRUIT
import com.karol.sezonnazdrowie.model.PREF_NOTIFICATION_TIME
import com.karol.sezonnazdrowie.model.PREF_NOTIFICATION_VEGETABLE
import com.karol.sezonnazdrowie.model.PREF_SEASON_END
import com.karol.sezonnazdrowie.model.PREF_SEASON_START
import com.karol.sezonnazdrowie.model.SnzAlarmManager
import com.karol.sezonnazdrowie.view.MainActivity
import com.karol.sezonnazdrowie.view.controls.TimePreference
import com.karol.sezonnazdrowie.view.controls.TimePreferenceDialogFragmentCompat
import org.threeten.bp.LocalTime
import org.threeten.bp.format.DateTimeFormatter

private const val DIALOG_TAG = "DIALOG"

class SettingsFragment : PreferenceFragmentCompat(),
    SharedPreferences.OnSharedPreferenceChangeListener, Preference.OnPreferenceClickListener {

    private val mainViewModel: MainViewModel by viewModels()

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.prefs, rootKey)

        setPreferenceSummary(preferenceManager.sharedPreferences, PREF_SEASON_START, true)
        setPreferenceSummary(preferenceManager.sharedPreferences, PREF_SEASON_END, false)
        setTimePreferenceSummary(preferenceManager.sharedPreferences, PREF_NOTIFICATION_TIME)

        findPreference<Preference>(PREF_NOTIFICATION_FRUIT)?.let {
            it.onPreferenceClickListener = this
        }
        findPreference<Preference>(PREF_NOTIFICATION_VEGETABLE)?.let {
            it.onPreferenceClickListener = this
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        mainViewModel.setActionBarTitle(getString(R.string.settings))
        super.onViewCreated(view, savedInstanceState)
    }

    override fun onResume() {
        super.onResume()
        preferenceManager.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
        if (mainViewModel.isSettingsItemChanged) { // TODO move logic to viewmodel
            SnzAlarmManager.startSetAlarmsTask(requireActivity(), mainViewModel.database)
        }
        mainViewModel.isSettingsItemChanged = false
    }

    override fun onPause() {
        preferenceManager.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
        super.onPause()
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        Log.d(TAG, "onSharedPreferenceChanged: key = $key")
        when (key) {
            PREF_SEASON_START -> setPreferenceSummary(sharedPreferences, key, true)
            PREF_SEASON_END -> setPreferenceSummary(sharedPreferences, key, false)
            PREF_NOTIFICATION_TIME -> setTimePreferenceSummary(sharedPreferences, key)
        }

        if (key in listOf(PREF_SEASON_START, PREF_SEASON_END, PREF_NOTIFICATION_TIME)) {
            SnzAlarmManager.startSetAlarmsTask(requireActivity(), mainViewModel.database)
        }
    }

    private fun setPreferenceSummary(
        sharedPreferences: SharedPreferences,
        key: String,
        seasonStart: Boolean
    ) {
        val summary = sharedPreferences.getStringSet(key, null)
            ?.asSequence()
            ?.map { SnzAlarmManager.PreNotiTimePeriod.fromPrefValue(it) }
            ?.sortedWith(Comparator { lhs, _ ->
                if (lhs == SnzAlarmManager.PreNotiTimePeriod.AtDay) {
                    -1
                } else {
                    if (lhs == SnzAlarmManager.PreNotiTimePeriod.WeekBefore) -1 else 1
                }
            })
            ?.map {
                when (it) {
                    SnzAlarmManager.PreNotiTimePeriod.AtDay -> {
                        if (seasonStart) R.string.at_the_start_day else R.string.at_the_end_day
                    }
                    SnzAlarmManager.PreNotiTimePeriod.WeekBefore -> R.string.week_before
                    SnzAlarmManager.PreNotiTimePeriod.MonthBefore -> R.string.month_before
                }
            }
            ?.map { context?.getString(it) }
            ?.joinToString()
            ?.takeIf {
                it.isNotEmpty()
            } ?: context?.getString(R.string.no_notification)
        findPreference<Preference>(key)?.summary = summary
    }

    private fun setTimePreferenceSummary(sharedPreferences: SharedPreferences, key: String) {
        val timeNano = sharedPreferences.getLong(key, 72000000000000L)
        val hourStr = LocalTime.ofNanoOfDay(timeNano)
            .format(DateTimeFormatter.ofPattern("HH:mm"))
        findPreference<Preference>(key)?.let { it.summary = hourStr }
    }

    override fun onPreferenceClick(preference: Preference): Boolean {
        val bundle = Bundle().apply {
            when (preference.key) {
                PREF_NOTIFICATION_FRUIT -> MainActivity.INTENT_WHAT_FRUITS
                PREF_NOTIFICATION_VEGETABLE -> MainActivity.INTENT_WHAT_VEGETABLES
                else -> return false
            }.let { putString(MainActivity.INTENT_WHAT, it) }
        }

        Navigation
            .findNavController(requireActivity(), R.id.nav_host_fragment)
            .navigate(R.id.settingsItemsFragment2, bundle)

        return true
    }

    override fun onDisplayPreferenceDialog(preference: Preference) {
        if (preference !is TimePreference) return super.onDisplayPreferenceDialog(preference)

        val dialogFragment = parentFragmentManager.findFragmentByTag(DIALOG_TAG)
        if (dialogFragment != null) {
            return
        }

        TimePreferenceDialogFragmentCompat.newInstance(preference.key, this)
            .show(parentFragmentManager, DIALOG_TAG)
    }

    companion object {

        private const val TAG = "SETTINGSFRAGMENT"
    }
}
